lib/src/any_map.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_docs)]
+
+use std::any::{Any, TypeId};
+use std::collections::HashMap;
+
+/// Type-safe map that stores objects of arbitrary types.
+///
+/// This allows extensions to store and retrieve their own types unknown to
+/// jj_lib safely.
+#[derive(Default)]
+pub struct AnyMap {
+    values: HashMap<TypeId, Box<dyn Any>>,
+}
+
+impl AnyMap {
+    /// Creates an empty AnyMap.
+    pub fn empty() -> Self {
+        Default::default()
+    }
+
+    /// Returns the specified type if it has already been inserted.
+    pub fn get<V: Any>(&self) -> Option<&V> {
+        self.values
+            .get(&TypeId::of::<V>())
+            .map(|v| v.downcast_ref::<V>().unwrap())
+    }
+
+    /// Inserts a new instance of the specified type.
+    ///
+    /// Requires that this type has not been inserted before.
```

Comment:
nit: since this requirement is specific to extension use case, it might be better to rename `AnyMap` to `ExtensionsMap` or something.